tests/instructorFiles.test.js

```diff
@@ -17,9 +18,9 @@ describe('File paths', () => {
     });
 
     it('works with valid absolute over relative paths', async () => {
-      assert.ok(contains('/PrairieLearn', 'tests'));
-      assert.ok(contains('/PrairieLearn', 'exampleCourse/questions'));
-      assert.ok(contains('/PrairieLearn/exampleCourse', 'exampleCourse/questions'));
+      assert.ok(contains(process.cwd(), 'tests'));
+      assert.ok(contains(process.cwd(), 'exampleCourse/questions'));
+      assert.ok(contains(path.join(process.cwd(), 'exampleCourse'), 'exampleCourse/questions'));
```

Comment:
Maybe I'm overthinking this, but if relative paths are interpreted relative to the current directory, that seems like undesirable behavior. For instance, say my working directory is `/foo/bar`. To check for something like `/bar/baz/biz`, I call `contains('/bar/baz', 'biz')`  which returns false.

Perhaps we want an implementation like this?

```ts
function contains(parentPath, childPath, includeSelf = true) {
  const absolutePath = path.resolve(parentPath, childPath);
  const relPath = path.relative(parentPath, absolutePath);
  if (!relPath) return includeSelf;
  return !(relPath.split(path.sep)[0] === '..' || path.isAbsolute(relPath));
}
```

This passes all tests except for the second test in "works with relative paths that are not contained":

```ts
assert.notOk(contains('PrairieLearn', 'tmp'));
```

That said, I wonder if that test is either wrong or meaningless. If `/foo/bar` can contain `baz`, why can't `bar` contain `baz`?

Reply:
In the case of that last test, is the intention that the items are './PrairieLearn' and './tmp', i.e. they are side by side?

Reply:
Maybe the issue is that `contains` is indeed not properly defined when the child path is relative and the parent path is absolute. I don't think this function is ever called with relative paths at all (I haven't checked thoroughly, though).

Reply:
Yeah, perhaps `contains` only has meaning when the `parentPath` is absolute. I think `childPath` can be relative, but I think we should always resolve it relative to `parentPath`.

Reply:
I checked all references of `contains`, and all of them have absolute paths in the `parentPath` and all are expected to have an absolute path in `childPath` (kinda, see below). So it is safe to say that, whatever we decide the behaviour should be if either `parentPath` or `childPath` is relative, it wouldn't affect the behaviour of any (well-behaved) call to this function.

About the kinda: there is a bit of an exception here. The "upload" feature in the file browser obtains the file path (for re-uploads of existing files) or the base directory (for uploads of new files) from the query. The default values for the query parameters in the hidden elements of the page are absolute addresses. I don't think this is exposing anything sensitive, since the course settings already expose the absolute path of the course directory anyway, and it is not actually directly susceptible to naming attacks, since the `contains` check in this function does what it needs to do to ensure that this upload is only allowed if the path is within the root of the specific context (e.g., question directory, assessment directory, etc.). However, if someone where to send a custom POST request that modifies these parameters to a relative path, that would be the only situation that I found where this `contains` would be called with a potentially relative path in `childPath`. So if we were to define a specific behaviour for the function, this kind of inadvertent "attack" would have a specific behaviour instead of an unexpected behaviour based on whatever the CWD is.

With that in mind, I think your suggestion of resolving relative paths based on the parent path makes sense, and would give the unconventional call above a well-defined behaviour.

I will change the code with your suggestion and update the tests accordingly.

Reply:
Maybe a weird edge case:
```
> contains('/', '..')
true
```
If the function is used everywhere with an absolute path for both parent and child, would it do to just raise an exception if that isn't the case going forward?

Reply:
> Maybe a weird edge case:
> ```
> > contains('/', '..')
> true
> ```

The parent of the root directory in POSIX systems is always the root itself. This test makes sense. If the third argument were false is expect this to be false as well.

> If the function is used everywhere with an absolute path for both parent and child, would it do to just raise an exception if that isn't the case going forward?

Not sure if it's worth it.